File: master-thesis-config-2/mtc-2-nodejs/algorithms/road-plan/logic/population.js
const algorithmsConfig = require('../../../helpers/algorithmsConfig');

class Population {
  roads;
  maxFit;

  constructor(roads) {
    this.roads = roads;
    this.maxFit = this.calculateMaxFit();
  }

  calculateMaxFit = () => {
    return this.roads.sort((a, b) => b.fitnessRatio - a.fitnessRatio)[0];
  };

  selection = () => {
    while (true) {
      let index = Math.floor(
        Math.random() * (algorithmsConfig.populationSize - 0 + 1) + 0
      );

      if (Math.random() < this.roads[index].fitnessRatio / this.maxFit)
        return new Road(this.roads[index].coordinates);
    }
  };

  generateNewPopulation = (size) => {
    let population = [];

    for (let i = 0; i < size; ++i) {
      let road = this.selection().crossing(this.selection());

      road.coordinates.forEach((coordinate) => {
        road = road.mutation();
      });

      population.push(road);
    }

    return new Population(population);
  };

  elite = (size) => {
    let best = [];
    let tmp = new Population(this.roads);

    for (let i = 0; i < size; ++i) {
      best.push(tmp.findBest());
      let roadsExceptBest = tmp.roads.filter((p) => !best.includes(p));
      tmp = new Population(roadsExceptBest);
    }

    return new Population(best);
  };

  findBest = () => {
    this.roads.forEach((road) => {
      if (road.fitnessRatio === this.maxFit) {
        return road;
      }
    });

    return null;
  };

  evolve = () => {
    const best = this.elite(algorithmsConfig.numberOfDominantsInNextGeneration);
    const newPopulation = this.generateNewPopulation(
      algorithmsConfig.populationSize -
        algorithmsConfig.numberOfDominantsInNextGeneration
    );

    return new Population([...best.roads, ...newPopulation.roads]);
  };
}

module.exports = Population;

File: master-thesis-config-3/mtc-3-dotnet/Domain/Algorithms/Population.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Domain.Algorithms.Models;

namespace Domain.Algorithms
{
    public class Population
    {
        public List<Road> p { get; private set; }
        public double maxFit { get; private set; }
        public static Random randomGenerator { get; set; }

        public Population(List<Road> l)
        {
            p = l;
            maxFit = this.calcMaxFit();
            randomGenerator = new Random();
        }

        public static Population randomized(Road t, int n)
        {
            List<Road> tmp = new List<Road>();

            for (int i = 0; i < n; ++i)
                tmp.Add( t.rearrange() );

            return new Population(tmp);
        }

        private double calcMaxFit() => p.Max( t => t.FitnessRatio );

        public Road select()
        {
            while (true)
            {
                int i = randomGenerator.Next(0, Config.populationSize);

                if (randomGenerator.NextDouble() < p[i].FitnessRatio / maxFit)
                    return new Road(p[i].Coordinates);
            }
        }

        public Population genNewPop(int n)
        {
            List<Road> p = new List<Road>();

            for (int i = 0; i < n; ++i)
            {
                Road t = select().performCrossing( select() );

                foreach (Coordinate c in t.Coordinates)
                    t = t.performMutation();

                p.Add(t);
            }

            return new Population(p);
        }

        public Population elite(int n)
        {
            List<Road> best = new List<Road>();
            Population tmp = new Population(p);

            for (int i = 0; i < n; ++i)
            {
                best.Add( tmp.findBest() );
                tmp = new Population( tmp.p.Except(best).ToList() );
            }

            return new Population(best);
        }

        public Road findBest()
        {
            foreach (Road t in this.p)
            {
                if (t.FitnessRatio == this.maxFit)
                    return t;
            }

            return null;
        }

        public Population evolve()
        {
            var best = elite(Config.numberOfDominantsInNextGeneration);
            var np = genNewPop(Config.populationSize - Config.numberOfDominantsInNextGeneration);
            return new Population( best.p.Concat(np.p).ToList() );
        }
    }
}